pkg/kubelet/pluginmanager/plugin_manager_test.go

```diff
@@ -140,13 +137,18 @@ func TestPluginRegistration(t *testing.T) {
 	p := pluginwatcher.NewTestExamplePlugin(pluginName, registerapi.DevicePlugin, socketPath, supportedVersions...)
 	require.NoError(t, p.Serve("v1beta1", "v1beta2"))
 
+	// Ensure example plugin is started before plugin manager, so that the socket is present
```

Comment:
/hold

If fsnotify is flaky and can miss events, then that seems like a product issue, where we need to design the plugin manager to be able to handle that?

Having the test start the plugin before the plugin manager seems like a workaround to get the test to pass, but masks potentially real problems. In a real system, the plugin is usually started after the plugin manager.

Reply:
I think what was racing here was not the filesystem and pluginmanager, but the observation of the plugin and the registration of the handler

in real life, we add the handlers that are sent the registration events before calling pluginManager.Run()

https://github.com/kubernetes/kubernetes/blob/ae9f76153f559c62bf1e7cf962d39168061dc6aa/pkg/kubelet/kubelet.go#L1301-L1307

would moving this above the goroutine where we call `pluginManager.Run(sourcesReady, stopChan)` be sufficient?
```
fakeHandler := newFakePluginHandler()
pluginManager.AddHandler(registerapi.DevicePlugin, fakeHandler)
```

Reply:
hmm, just saw in the PR description:

> Please notice when the test fails, the plugin watcher could neither find any sockets in the socket dir nor receive "CREATE" event later. I suspect fsnotify is not extremely reliable and might miss events in some cases.
> 
> As a result, I decide to make sure the example plugin server is started before plugin manager, so that the socket always exists when the plugin manager is started. I have run the test 100 times locally after this change, they all pass.

that's troubling, and I agree with @msau42 that if that is the issue we need a real fix, not to mask it in a test

Reply:
I agree, let me try creating a minimal reproducible example first.